src/rust/src/x509.rs

```diff
@@ -544,6 +547,441 @@ fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<
     })
 }
 
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+struct RawCsr<'a> {
+    csr_info: CertificationRequestInfo<'a>,
+    signature_alg: AlgorithmIdentifier<'a>,
+    signature: asn1::BitString<'a>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+struct CertificationRequestInfo<'a> {
+    version: u8,
+    subject: Name<'a>,
+    spki: asn1::Sequence<'a>,
+    #[implicit(0)]
+    // TODO: this needs to be non-optional, but we need to fix rust-asn1 for that
+    attributes: Option<asn1::SetOf<'a, Attribute<'a>>>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+struct Attribute<'a> {
+    type_id: asn1::ObjectIdentifier<'a>,
+    values: asn1::SetOf<'a, asn1::Tlv<'a>>,
+}
+
+fn check_attribute_length<'a>(values: asn1::SetOf<'a, asn1::Tlv<'a>>) -> Result<(), PyAsn1Error> {
+    if values.count() > 1 {
+        Err(PyAsn1Error::from(pyo3::exceptions::PyValueError::new_err(
+            "Only single-valued attributes are supported",
+        )))
+    } else {
+        Ok(())
+    }
+}
+
+impl CertificationRequestInfo<'_> {
+    fn get_extension_attribute<'a>(&'a self) -> Result<Option<Extensions<'a>>, PyAsn1Error> {
+        // TODO: no more unwrap once attributes isn't an optional. Seems like we could
+        // eliminate a bunch of clones here too
+        for attribute in self.attributes.clone().unwrap() {
+            if attribute.type_id == *EXTENSION_REQUEST || attribute.type_id == *MS_EXTENSION_REQUEST
+            {
+                check_attribute_length(attribute.values.clone())?;
+                if let Some(val) = attribute.values.clone().next() {
+                    return Ok(Some(asn1::parse_single::<Extensions<'a>>(val.full_data())?));
+                }
+            }
+        }
+        Ok(None)
+    }
+}
+
+#[ouroboros::self_referencing]
+struct OwnedRawCsr {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    value: RawCsr<'this>,
+}
+
+#[pyo3::prelude::pyclass]
+struct CertificateSigningRequest {
+    raw: OwnedRawCsr,
+    cached_extensions: Option<pyo3::PyObject>,
+}
+
+#[pyo3::prelude::pyproto]
+impl pyo3::class::basic::PyObjectProtocol for CertificateSigningRequest {
```

Comment:
There's a whole bunch of names that can be shortened: none of the `::class::` are required nor are the `::pycell::`